include/pybind11/detail/init.h

```diff
@@ -61,24 +61,53 @@ constexpr bool is_alias(void *) {
     return false;
 }
 
-// Constructs and returns a new object; if the given arguments don't map to a constructor, we fall
+// Constructs a new object inplace; if the given arguments don't map to a constructor, we fall
 // back to brace aggregate initiailization so that for aggregate initialization can be used with
 // py::init, e.g.  `py::init<int, int>` to initialize a `struct T { int a; int b; }`.  For
 // non-aggregate types, we need to use an ordinary T(...) constructor (invoking as `T{...}` usually
 // works, but will not do the expected thing when `T` has an `initializer_list<T>` constructor).
-template <typename Class,
-          typename... Args,
-          detail::enable_if_t<std::is_constructible<Class, Args...>::value, int> = 0>
-inline Class *construct_or_initialize(Args &&...args) {
-    return new Class(std::forward<Args>(args)...);
+template <
+    typename Class,
+    bool Preallocate,
+    typename... Args,
+    detail::enable_if_t<std::is_constructible<Class, Args...>::value && !Preallocate, int> = 0>
+inline void construct_or_initialize(value_and_holder &v_h, Args &&...args) {
+    v_h.value_ptr() = new Class(std::forward<Args>(args)...);
 }
-template <typename Class,
-          typename... Args,
-          detail::enable_if_t<!std::is_constructible<Class, Args...>::value, int> = 0>
-inline Class *construct_or_initialize(Args &&...args) {
-    return new Class{std::forward<Args>(args)...};
+template <
+    typename Class,
+    bool Preallocate,
+    typename... Args,
+    detail::enable_if_t<!std::is_constructible<Class, Args...>::value && !Preallocate, int> = 0>
+inline void construct_or_initialize(value_and_holder &v_h, Args &&...args) {
+    v_h.value_ptr() = new Class{std::forward<Args>(args)...};
+}
+// The preallocated variants are performing memory allocation and registration before actually
+// calling the constructor to allow casting the C++ pointer to its Python counterpart.
+template <
+    typename Class,
+    bool Preallocate,
+    typename... Args,
+    detail::enable_if_t<std::is_constructible<Class, Args...>::value && Preallocate, int> = 0>
+inline void construct_or_initialize(value_and_holder &v_h, Args &&...args) {
+    v_h.value_ptr() = malloc(sizeof(Class));
```

Comment:
I am pretty sure we only 'Pymem_alloc*' apis for portability reasons.

Reply:
+1

I need to find time to look at this carefully, but a quick search for "free" didn't match anything in the diff.

@adriendelsalle did you look already how the memory is deallocated? We want to be sure it is compatible with the allocation mechanism.

Purely from memory and hand-waving: I believe for old-style `__init__` the current code can wind up with a `delete` that has no matching `new` (some other allocation mechanism is used). If you happen to stumble over that, too, adding a comment would be great. (I wanted to do that but never got back to it.)

Reply:
Awesome to have your comments!
I'm using the lib for years but only started to go in the details few days ago..so I'm not really comfortable with alloc/dealloc.
Is there some dev or design/architecture doc for that? I'll do my best to digest that tomorrow!

Reply:
> Is there some dev or design/architecture doc for that?
Not to my knowledge.
At least I found it necessary to learn the hard way, experimenting a lot with the code.
Feel free to ask questions here. Maybe we can learn together.

Reply:
Seeing the only now, the valgrind failure seems to be spot-on:

https://github.com/pybind/pybind11/runs/7732180285?check_suite_focus=true

> ==7330== Mismatched free() / delete / delete []

Reply:
I just pushed a change to replace `malloc` with `new` allocation to be fully consistent with `delete` used in `dealloc`. It should be then very similar to what's done in the allocation+construction overload of `construct_or_initialize`.

> I am pretty sure we only 'Pymem_alloc*' apis for portability reasons.

I don't really get the point, could you please elaborate a bit? I don't see why allocation and construction of a C++ object should use some `Py*` alloc.

Here is my understanding of how the changes integrate with the implem:
- passing `py::preallocate()` as an extra argument of a new style constructor will preallocate and register just before actually calling the constructor of the class
- if the call is fine, the holder is still not registered making this piece of code to be called (only to init the holder then):
https://github.com/pybind/pybind11/blob/bbb89da152a0e4f63a4fecc0d4db44f0ad855fdf/include/pybind11/pybind11.h#L1127-L1130
- if the call is not successful (because of cast error AFAIU), the next overload try (if any) will deallocate the memory if already allocated
https://github.com/pybind/pybind11/blob/bbb89da152a0e4f63a4fecc0d4db44f0ad855fdf/include/pybind11/pybind11.h#L754-L758
  - maybe it would worth to update the comment because now it's not *only* in case of old-style `__init__` that the memory is already allocated?
  - I don't really understand the comment:
https://github.com/pybind/pybind11/blob/bbb89da152a0e4f63a4fecc0d4db44f0ad855fdf/include/pybind11/pybind11.h#L440-L442
since it looks like to me the opposite: a new-style `__init__` takes `value_and_holder` as `self`, isn't it?
https://github.com/pybind/pybind11/blob/bbb89da152a0e4f63a4fecc0d4db44f0ad855fdf/include/pybind11/pybind11.h#L761-L762